### PR TITLE
Integrating ECE to shortcode checkout

### LINKED ISSUE #3404: [ECE] Implement ECE for payment processing on shortcode checkout page
### Description
Process payment with the ECE button on the shortcode checkout page.

### Acceptance Criteria
- ECE buttons are rendered on the shortcode checkout page.
- You can successfully process payment with the ECE button from the shortcode checkout page.
- Both success and failure cases are handled including necessary notices.

### PR DESCRIPTION
**Pull request by wjrosa**
Fixes #3404
Doc: https://docs.stripe.com/elements/express-checkout-element

## Changes proposed in this Pull Request:

This PR introduces the new Stripe Express Checkout Element to our shortcode checkout, replacing the current PRB version.

This integration is heavily based on [WCPay's](https://github.com/Automattic/woocommerce-payments/), with just some adjustments to adapt the code to our needs and standards.

## Testing instructions

- Apply this PR to your test environment (`add/ece-for-shortcode-checkout`)
- Make sure it is publicly accessible. You can use [ngrok](https://dashboard.ngrok.com/) for this (or a JN website)
- Enable the ECE flag. You can hardcode `is_stripe_ece_enabled` to `true`
- As a merchant, connect your Stripe account and create a shortcode checkout page
- As a customer, add any product to your cart
- Go to the shortcode checkout page
- Confirm that the new ECE buttons are loaded:
Screenshot 2024-09-19 at 14 26 50

- Confirm that you can place an order successfully

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

cc @Mayisha

### CODE REVIEW

File: client/entrypoints/express-checkout/index.js [hunk 1]
```diff
@@ -1 +1,384 @@
...
+									amount: response.total.amount,
+								} );
+							} else {
+								wcStripeECE.reInitExpressCheckoutElement(
+									response
+								);
+							}
+						} )
+						.catch( () => {
+							wcStripeECE.hide();
```
Comment (mattallan) [inline client/entrypoints/express-checkout/index.js:380]: Is this `hide()` function missing?

File: includes/payment-methods/class-wc-stripe-express-checkout-element.php [hunk 2]
```diff
@@ -324,7 +328,7 @@ public function display_express_checkout_button_html() {
 		}
 
 		?>
-		<div id="wc-stripe-express-checkout-wrapper" style="margin-top: 1em;clear:both;display:none;">
+		<div id="wc-stripe-express-checkout-wrapper" style="margin-top: 1em;clear:both;">
```
Comment (mattallan) [inline includes/payment-methods/class-wc-stripe-express-checkout-element.php]: Curious why you removed the `displey:none;` here?
I think the default should always be display none and then show the elements via JS.

WooPayments just made this change in a recent fix as well that we might need to bring over to Stripe: https://github.com/Automattic/woocommerce-payments/commit/8f2b1ac66610479282b375078ba00b5aa4d1dfe9

File: includes/class-wc-stripe-feature-flags.php [hunk 3]
```diff
@@ -14,7 +14,8 @@ class WC_Stripe_Feature_Flags {
 	 * @return bool
 	 */
 	public static function is_stripe_ece_enabled() {
-		return 'yes' === get_option( self::ECE_FEATURE_FLAG_NAME, 'no' );
+		// @todo revert this
```
Comment (mattallan) [inline includes/class-wc-stripe-feature-flags.php]: Just adding a comment here to make sure we remember to revert this 😅

File: client/entrypoints/express-checkout/utils/normalize.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		shipping_address_2: shipping?.address?.line2 ?? '',
+		shipping_city: shipping?.address?.city ?? '',
+		shipping_state: shipping?.address?.state ?? '',
+		shipping_postcode: shipping?.address?.postal_code ?? '',
+		shipping_method: [ event?.shippingRate?.id ?? null ],
+		order_comments: '',
+		payment_method: 'woocommerce_payments',
+		ship_to_different_address: 1,
+		terms: 1,
+		'w-stripe-payment-method': paymentMethodId,
```
Comment (mattallan) [inline client/entrypoints/express-checkout/utils/normalize.js]:
```suggestion
		'wc-stripe-payment-method': paymentMethodId,
```

File: client/entrypoints/express-checkout/index.js [hunk 5]
```diff
@@ -1 +1,384 @@
...
+		// A promise-based interface to jQuery.post.
+		( url, args ) => {
+			return new Promise( ( resolve, reject ) => {
+				jQuery.post( url, args ).then( resolve ).fail( reject );
+			} );
+		}
+	);
+	const elements = api.getStripe().elements( {
+		mode: 'payment',
+		amount: 1099,
```
Comment (mattallan) [inline client/entrypoints/express-checkout/index.js]: Also making a note of some dummy data used here.

**Review by mattallan — COMMENTED** [review]
Thanks for kicking these changes off Wesley, I've just taken a quick code review and left some notes.

I will base my branch off this PR to start looking at implementing ECE on the product page

File: client/entrypoints/express-checkout/tracking.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+import { recordUserEvent } from 'tracks';
```
Comment (mattallan) [inline client/entrypoints/express-checkout/tracking.js]: Should be using `import { recordEvent } from 'wcstripe/tracking';` here.

File: client/entrypoints/express-checkout/utils/normalize.js (hunk 4 above)
Comment (mattallan) [reply client/entrypoints/express-checkout/utils/normalize.js]: Fixed in [0af7c3e](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/0af7c3ee33a2b6aff8623927caf7870ae1fbf016)

File: client/entrypoints/express-checkout/tracking.js (hunk 6 above)
Comment (mattallan) [reply client/entrypoints/express-checkout/tracking.js]: Fixed in [8956e8e](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/8956e8e9ebd2217c7998e16d3cc6c3286bf40296)

File: .eslintrc.js [hunk 7]
```diff
@@ -18,7 +18,6 @@ module.exports = {
 		node: true,
 	},
 	rules: {
-		'require-await': 'error',
```
Comment (mattallan) [inline .eslintrc.js]: Thoughts on removing this @wjrosa ?

This rule was causing issues with using code like:

```
eceButton.on( 'shippingaddresschange', async ( event ) =>
	shippingAddressChangeHandler( api, event, elements )
);
```

Even though the await was within `shippingAddressChangeHandler`. I think this rule might be too restrictive so I decided to remove it.

Happy to discuss and go with another alternative :)

File: client/entrypoints/express-checkout/index.js (hunk 1 above)
Comment (mattallan) [reply client/entrypoints/express-checkout/index.js:380]: Added in acc8e9222161b5052b17458107f80ba31692edee

**Comment by mattallan** [comment]
Hey @wjrosa 👋 I hope this hasn't conflicted with any changes you have locally, but I've just pushed up a few changes [acc8e92](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/acc8e9222161b5052b17458107f80ba31692edee) adds the missing ECE AJAX calls that were being referenced in the event-handlers file. I later pushed up [9a40a5d](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/9a40a5d68f5c51359d3af60e0bc7dcb7dfe1e279) to use the `getExpressCheckoutData` function.

Not sure if you wanted these functions within `client/api` or somewhere else.
I put them in client/api to follow WooPayments but happy for it to go wherever :)

File: .eslintrc.js (hunk 7 above)
Comment (james-allan) [reply .eslintrc.js]: I'm not 100% sure (this isn't my strong suit) but I think another alternative is to remove the async keyword. 

eg 

```js
eceButton.on( 'shippingaddresschange', ( event ) =>
	shippingAddressChangeHandler( api, event, elements )
);
``` 

The linting error is being thrown because the callback function is defined as async but that function doesn't use the `await` keyword directly. There's really 3 options to appease the linting error. 

1. Remove this rule. Though I'm not sure if that's a good idea. 
2. Remove the async keyword. (See example above).
3. Add an await. See example below.

```js
eceButton.on( 'shippingratechange', async ( event ) =>
	await shippingRateChangeHandler( api, event, elements )
);
```

Comment (wjrosa) [reply .eslintrc.js]: Adding the `await` keyword still triggers the error `async functions are only available in ES8` on the `async` keyword. What if we also mirror our ESlint rules with WCPay? I see they have some different versions for the ESLint packages, and they use the `override` key on .eslintrc.js

File: includes/payment-methods/class-wc-stripe-express-checkout-element.php (hunk 2 above)
Comment (wjrosa) [reply includes/payment-methods/class-wc-stripe-express-checkout-element.php]: Good catch, Matt! I reverted this change here and followed the implementation you liked. It seems to be working fine now 👍

**Comment by wjrosa** [comment]
> Hey @wjrosa 👋 I hope this hasn't conflicted with any changes you have locally, but I've just pushed up a few changes [acc8e92](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/acc8e9222161b5052b17458107f80ba31692edee) adds the missing ECE AJAX calls that were being referenced in the event-handlers file. I later pushed up [9a40a5d](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/9a40a5d68f5c51359d3af60e0bc7dcb7dfe1e279) to use the `getExpressCheckoutData` function.
> 
> Not sure if you wanted these functions within `client/api` or somewhere else. I put them in client/api to follow WooPayments but happy for it to go wherever :)

Thank you so much for working on this! I love collaborative work ❤️ . This is now on a much better place, and it is working nice for me (still need to test the whole flow). I have also added some missing methods to the main JS file, renamed the main `div` (and fixed the references to it), and changed some of the code to match our ESlint rules.

File: .eslintrc.js (hunk 7 above)
Comment (wjrosa) [reply .eslintrc.js]: I have updated our ESLint version (and the min requirement on .jshintrc) in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/bc581aa7fe0a73f2c698d39a1082ca87497c749a . Tests are still working and I could not see any issues with it. Let me know if you think that's a good move or if it is better to revert it 👍

**Comment by mattallan** [comment]
Hey @wjrosa 👋 Glad you don't mind the collaboration!

While continuing with the product page side of things, I found some issues calculating and updating shipping options, that would also be impacting the ECEs on checkout page (and cart pages).

I've just pushed up [bd7dacd](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/bd7dacdccedc6e9632c92c8ea698bfee60237039) to fix these.

I'm trying to avoid as many conflicts as possible with both of us working on product and checkout pages, so I'll continue to push up any small changes I make that will also fix issues for you 😅

**Comment by mattallan** [comment]
Hey again @wjrosa I've just pushed up a couple of small commits.
- [e49eb3a](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/e49eb3a4d6b40e34d2cd3ba2c90a77c5fa7d55b2) - This commit basically fixed everything for me and now I'm successfully processing google pay payments 😄 I suspect this might fix a lot of things for you on the checkout page as well.

Not sure if you were running into the same issues but when I was trying to purchase from the product page, I kept getting a "Payment processing failed. Please retry." error. Turns out we just needed to set the "use deferred intent" flag

props @james-allan for the suggestion! 🙏

**Comment by wjrosa** [comment]
> Hey again @wjrosa I've just pushed up a couple of small commits.
> 
> * [e49eb3a](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/e49eb3a4d6b40e34d2cd3ba2c90a77c5fa7d55b2) - This commit basically fixed everything for me and now I'm successfully processing google pay payments 😄 I suspect this might fix a lot of things for you on the checkout page as well.
> 
> Not sure if you were running into the same issues but when I was trying to purchase from the product page, I kept getting a "Payment processing failed. Please retry." error. Turns out we just needed to set the "use deferred intent" flag
> 
> props @james-allan for the suggestion! 🙏

Thank you so much for all the hard work, Matt! Yes, I was having the same issue and now it is working fine for me as well. Can you elaborate a bit about this `use deferred intent` flag? Not aware of it.

Also, I am still having an issue that we need to fix. When the payment fails, I get this error on console:
```
Uncaught (in promise) IntegrationError: Unexpected call to paymentFailed(). Ensure you are either submitting a payment or calling paymentFailed() once per expressCheckout Element confirm event.
```
It appears to be related to `paymentFailed` being called multiple times. But I did some debugging and it is not 🤔 . I will keep investigating. Do you know how to force failures with GPay? I have only some pre-defined cards on my test wallet, and there's no option to include new ones.

File: .eslintrc.js (hunk 7 above)
Comment (wjrosa) [reply .eslintrc.js]: I have reverted this rule removal as I fixed all the warnings in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/1e080897c523f5776b7523a96d4bd78c9d7890f4

File: .jshintrc [hunk 8]
```diff
@@ -1,5 +1,5 @@
 {
-	"esversion": 6,
+	"esversion": 11,
```
Comment (wjrosa) [inline .jshintrc:2]: I did some research, and this actually means [the minimum ECMAScript-supported version](https://jshint.com/docs/options/#esversion). The [11th version was introduced in June 2020](https://262.ecma-international.org/11.0/index.html), and it supports optional property chaining (which we already use in many places of the code).

File: includes/class-wc-stripe-feature-flags.php (hunk 3 above)
Comment (wjrosa) [reply includes/class-wc-stripe-feature-flags.php]: Done! Reverted it

File: .jshintrc [hunk 9]
```diff
@@ -13,6 +13,9 @@
 
 	"browser": true,
 
+	"mocha": true,
```
Comment (wjrosa) [inline .jshintrc:16]: This fixes undefined functions in unit tests

**Comment by james-allan** [comment]
@wjrosa just a heads up that I pushed 2 commits with uncontroversial changes - updating `woocommerce-payments` ([a5b19e4](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/a5b19e4897f0300e88274215caeaba7420d71f29)) i18n text domains and a minor typo in a function block comment ([7a53b7c](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/7a53b7c9ae1dd2968145adcd4a7b86b7fb0d02b0)).

**Comment by james-allan** [comment]
@wjrosa I'm getting this error when attempting to place the order using Google Pay.

Screenshot 2024-09-23 at 12 58 26 pm

From the quick glance, it looks like its based on the code from WooPayments which has a whole fraud prevention class that generates a unique token for the user and stores it in a session variable and then injects it into the JS. That's all missing in the Stripe plugin. 

Given the size of that, it's probably not something we want to handle as part of this PR but in that case we'll need to remove references to it so this PR can place orders if that's the base you want to work off.

**Comment by Mayisha** [comment]
@wjrosa I also see the [same error](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436#issuecomment-2367153970) James posted above. When I remove all the uses of `wcStripeFraudPreventionToken` and try to pay with Google Pay, I get the following error-

Screenshot 2024-09-23 at 3 41 31 PM

**Comment by Mayisha** [comment]
For blocks and shortcode implementation we have the same functions in different places and with different names. 
I have moved all the common functions (utils, event handlers, normalizers etc.) to a common folder `client/express-checkout` in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3458. 

Codes are copied over from this PR and https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3439.

**Comment by wjrosa** [comment]
Thanks, @Mayisha ! Can you and @james-allan test this again now that I have rebased it with your newly created PR?

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php [hunk 10]
```diff
@@ -615,23 +633,22 @@ public function get_shipping_options( $shipping_address, $itemized_display_items
...
 					}
 
-					foreach ( $package['rates'] as $key => $rate ) {
+					foreach ( $package['rates'] as $rate ) {
 						if ( in_array( $rate->id, $shipping_rate_ids, true ) ) {
-							// The express checkout will try to load indefinitely if there are duplicate shipping
-							// option IDs.
-							throw new Exception( __( 'Unable to provide shipping options for express checkout.', 'woocommerce-gateway-stripe' ) );
+							// The Payment Requests will try to load indefinitely if there are duplicate shipping option IDs.
+							throw new Exception( __( 'Unable to provide shipping options for Payment Requests.', 'woocommerce-gateway-stripe' ) );
```
Comment (Mayisha) [inline includes/payment-methods/class-wc-stripe-express-checkout-helper.php:644]: QQ - Is the wording of this message and the comment above intentionally changed from `express checkout` to `Payment requests`?

File: client/entrypoints/express-checkout/index.js [hunk 11]
```diff
@@ -1 +1,472 @@
...
+			} else if ( getExpressCheckoutData( 'is_product_page' ) ) {
+				// Product page specific initialization.
+			} else {
+				let requestPhone = false;
+				if ( getExpressCheckoutData( 'checkout' ).needs_payer_phone ) {
+					requestPhone = getExpressCheckoutData( 'checkout' )
+						.needs_payer_phone;
+				}
+				// Cart and Checkout page specific initialization.
+				// TODO: Use real cart data.
```
Comment (Mayisha) [inline client/entrypoints/express-checkout/index.js]: Will this be addressed in a separate PR?

Comment (wjrosa) [reply client/entrypoints/express-checkout/index.js]: Just did it on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436/commits/e321b5db0591235df8f6391195ad7739a3629421

**Comment by wjrosa** [comment]
@Mayisha, I have fixed the issue you reported on Slack. Some parameters were missing from the fake cart data provided. Now, it is working fine for me (I hope it fixes it for you folks, too).

File: includes/payment-methods/class-wc-stripe-express-checkout-helper.php (hunk 10 above)
Comment (wjrosa) [reply includes/payment-methods/class-wc-stripe-express-checkout-helper.php:644]: I think so, to make sense with the class name. This specific line was changed by @mattallan

**Comment by james-allan** [comment]
This looks good to me. I tested a couple of different general scenarios. 


- [x] Basic purchases as a logged in user and logged out (as a guest). 
- [x] Subscription purchases. 
- [x] Subscription renewals. 
- [x] Free trial subscriptions. (Not available - see `is_invalid_subscription_product()`)
- [x] Errors - I'm not sure if there's good way to test failures, however, I forced errors by invaliding the payment request.

Screenshot 2024-09-25 at 3 27 05 pm

- [x] I changed the checkout button settings in the plugin and confirmed they are reflected on the checkout
 
Screenshot 2024-09-25 at 3 28 53 pm

- [ ] 🔴  I checkouts with a coupon applied. This failed and the checkout remained blocked.

Screenshot 2024-09-25 at 3 37 32 pm

- [x] 🟢 I tested various shipping setups and confirmed that the order total in the model and the Stripe Dashboard was correct.
    - I chose a shipping method, changed it via the Google Pay pop up. See example video below. 

https://github.com/user-attachments/assets/f09ed8fb-2bfa-4545-98e3-48c258a7e35c\

> I checkouts with a coupon applied. This failed and the checkout remained blocked.

I'm happy if we want to handle this separately given the large size of this PR and the other work depending on this PR being merged.

@Mayisha I'm not sure if there are other specific tests you had you want us to check as part of this feature. Let me know if there are specific flows we should check etc. :)

**Review by Mayisha — APPROVED** [review]
Thanks @wjrosa for the implementation. I have tested with Google Pay and everything is working on the shortcode checkout page.

✅ Purchasing simple product, digital product, simple subscription work.
✅ Checkout with different shipping methods (free, flat rate) works. Tested by changing the shipping method both from checkout page and Google Pay modal. 
✅ I can not reproduce [the issue](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436#issuecomment-2367718372) related to shipping address anymore.

Thanks @james-allan for testing other scenarios. I have created separate issues for the coupon and payment method title errors.
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3469
- https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3471
